Docker image with prebuilt pyodide

MischaPanch opened the pull request:
This PR addresses the [issue 656](https://github.com/iodide-project/pyodide/issues/656) about having a docker image with a pre-built version of pyodide. Using this prebuilt image radically reduces the time required for custom builds, e.g. for transpiling one's own python package to javascript (this was my use case)

The image will be built and published through a manually triggered github workflow, you will need to add secrets for authenticating to dockerhub in your github account.

If you are happy with the changes here, I would also extend the docu explaining how the new image can be used

rth commented:
Thanks @MischaPanch , overall looks good!

A few comments,
 - the current version of pyodide [is 0.15.0](https://github.com/iodide-project/pyodide/tags), the next one will be 0.16.0, so we should tag these image as `0.16.0b1` or something similar (i.e. pre-release).
 - the version of `pyodide-env` docker image, currently evolves independently of the pyodide releases. For instance, we won't necessary push a new `pyodide-env` image when we make the next release, if none of the build dependencies change. And inversely we may have to update and tag this image in between releases (e.g. updating Debian version etc) since it's used in CI. Maybe we should switch it to [CalVer](https://calver.org/) or something else to avoid the confusion with pyodide release numbers.

Also I just realized that I don't have enough permissions to set secrets in this repo, let me look into that and I'll get back in a couple of days. 

(As a side note, I find it really strange that they need QEMU to build docker image. I know it's in the official docs, but still surprising. Maybe let's add a link to the [docs](https://github.com/marketplace/actions/build-and-push-docker-images) in a comment in the pipeline setup.)

MischaPanch commented:
> Thanks @MischaPanch , overall looks good!
> 
> A few comments,
> 
> * the current version of pyodide [is 0.15.0](https://github.com/iodide-project/pyodide/tags), the next one will be 0.16.0, so we should tag these image as `0.16.0b1` or something similar (i.e. pre-release).
> * the version of `pyodide-env` docker image, currently evolves independently of the pyodide releases. For instance, we won't necessary push a new `pyodide-env` image when we make the next release, if none of the build dependencies change. And inversely we may have to update and tag this image in between releases (e.g. updating Debian version etc) since it's used in CI. Maybe we should switch it to [CalVer](https://calver.org/) or something else to avoid the confusion with pyodide release numbers.
> 
> Also I just realized that I don't have enough permissions to set secrets in this repo, let me look into that and I'll get back in a couple of days.
> 
> (As a side note, I find it really strange that they need QEMU to build docker image. I know it's in the official docs, but still surprising. Maybe let's add a link to the [docs](https://github.com/marketplace/actions/build-and-push-docker-images) in a comment in the pipeline setup.)


Thank you for the comments, @rth ! Since I don't want to mess with your release mechanism, I the version of the pre-built image is now set manually when the workflow is triggered. This also facilitates building the pre-built image for older versions. I actually need the 0.15.1 version for my own project, since there is some kind of regression bug in 0.16.1 (I will try to open an issue on that in the next days)

I also pimped your run_docker script a bit at the cost that it uses bash instead of sh. I hope that's ok

rth commented:
Thanks @MischaPanch! I tried to run the `run_docker` script but I'm getting the following errors,
```
$ bash ./run_docker 
./run_docker: line 66: GID: unbound variable
$ ./run_docker -p 8000
./run_docker: line 49: PYODIDE_SYSTEM_PORT: unbound variable
$ ./run_docker --pre-built
./run_docker: line 37: PYODIDE_DOCKER_IMAGE: unbound variable
$ ./run_docker --pre-built
```
my default shell is zsh, but should be able to find bash with the first line of the script.

MischaPanch commented:
Yes, sorry, I forgot to remove the -u option. It is not necessary since there are defaults for all env variables. I need to have checks that involve unbound variables since I wanted the script to be backwards compatible

MischaPanch commented:
In fact, the set -u was not there in your original script and I shouldn't have used it in first place. The script now involves the same unbound variables that it had before

rth commented:
Thanks @MischaPanch ! Generally LGTM, I think we can merge and push the image manually for now.

Just one last thing. I wrote a short test for the argument parsing in this script, could you please add it,
```py
# Add to test/pyodide_build/test_run_docker.py
from pathlib import Path
import subprocess

BASE_DIR = Path(__file__).parents[2]


def test_run_docker_script():
    res = subprocess.run(
        ["bash", str(BASE_DIR / "run_docker"), "--help"],
        check=False,
        stdout=subprocess.PIPE,
        stderr=subprocess.PIPE,
    )

    assert "Usage:\n  run_docker" in res.stderr.decode("utf-8")

    res = subprocess.run(
        ["bash", str(BASE_DIR / "run_docker"), "--invalid-param"],
        check=False,
        stdout=subprocess.PIPE,
        stderr=subprocess.PIPE,
    )
    assert res.returncode > 0
    assert "Unknown option --invalid-param" in res.stdout.decode("utf-8")
```

A somewhat inconsistent behavior is that `./run_docker --help` returns a non 0 status and write to stderr, while `./run_docker --invalid-param` writes to stdout..

MischaPanch commented:
Nice catch on the inconsistencies in the output! I fixed it and added your test (after modifying it accordingly). I would be happy if you could build the images manually, then I don't have to use my custom image anymore.

A side note: I will soon open source a cookiecutter template for pyodide based frontend projects (with support for react and possibly angular). It generates a project where python code is packaged, transpiled within CI (with the new pre-built image), included into frontend code with a web-worker, loaded in the background and cached (such that the large files are loaded in a separate thread and don't bother the user). Do you think this template would be worth mentioning somewhere in your docs? It certainly would have helped me to set up my project

rth commented:
Thank you @MischaPanch ! Trying to build it locally now. I think we can also add,
```diff
--- a/Dockerfile-prebuilt
+++ b/Dockerfile-prebuilt
@@ -4,6 +4,8 @@ FROM iodide/pyodide-env:${VERSION}
 
 USER root
 
+ENV EMSDK_NUM_CORES=4 EMCC_CORES=4 PYODIDE_JOBS=4
+
 COPY . pyodide
 
 RUN cd pyodide && make
```
to make the build a bit faster, assuming a reasonably modern CPU.

>  side note: I will soon open source a cookiecutter template for pyodide based frontend projects

Yes, that would be great. We should create

rth commented on Dockerfile-prebuilt:
```diff
@@ -0,0 +1,14 @@
+# this deletes the build results such that the resulting image can still be used for building pyodide
+# from source (including partial and customized builds). Due to the previous run of make, builds
+# executed with this image will be much faster than the ones executed with pyodide-env
+RUN rm -r build
```
Are you sure this is needed? We could just remove the `build/package.json` and it should work I think. That way for packages depending on an existing dependency (say numpy) it wouldn't have to be rebuilt.

MischaPanch replied:
I think it depends on how the image is supposed to be used. For example, I used it to build my package and copy everything from the build dir to the dir with frontend code. Generally, by performing a partial build after setting the PYODIDE_PACKAGES env var, I wouldn't expect any other packages in the build directory. Even if the unnecessary packages are filtered out in a subsequent build step (e.g. with npm build), it would be natural to have the contents of the build dir as artifacts (say in a ci job) and they should not be larger than required. I think partial builds will be the most common practice (at least for the purpose of frontend applications), since the entire pyodide stack is huge in size.

Also, rebuilding with this image is pretty fast (even numpy). But maybe I'm missing something here, I am not very familiar with your build process.

rth replied:
Yes, you are right we can remove it. Rebuilding is fast as all the package build folder (`packages/*/build`) are still there).

MischaPanch commented:
Done, I also made a comment in the docu. Two questions though - doesn't the build currently use all available CPUs by default? Also, these vars should probably also be part of the pyodide-env image. I couldn't find the dockerfile for it (I don't think it is the one here used for circleci, or is it?). Could you add the vars to it, in case it is public?

> Thank you @MischaPanch ! Trying to build it locally now. I think we can also add,
> 
> ```diff
> --- a/Dockerfile-prebuilt
> +++ b/Dockerfile-prebuilt
> @@ -4,6 +4,8 @@ FROM iodide/pyodide-env:${VERSION}
>  
>  USER root
>  
> +ENV EMSDK_NUM_CORES=4 EMCC_CORES=4 PYODIDE_JOBS=4
> +
>  COPY . pyodide
>  
>  RUN cd pyodide && make
> ```

Awesome, thanks @rth ! I will open a PR with the related projects page when I open source the template then.

> to make the build a bit faster, assuming a reasonably modern CPU.
> 
> > side note: I will soon open source a cookiecutter template for pyodide based frontend projects
> 
> Yes, that would be great. We could create a "related projects" page in the documentation.

rth commented:
The image [`iodide/pyodide:0.16.0b1`](https://hub.docker.com/repository/docker/iodide/pyodide) should now be available.

I had to make 2 changes with respect to this PR to make it work,
<details>

```diff
diff --git a/.dockerignore b/.dockerignore
index 0ab130a..5e9654a 100644
--- a/.dockerignore
+++ b/.dockerignore
@@ -1,4 +1,4 @@
 emsdk/emsdk/
 cpython/installs/
 cpython/build/
-packages/
+packages/*/build/
diff --git a/Dockerfile-prebuilt b/Dockerfile-prebuilt
index 5418aef..6392d62 100644
--- a/Dockerfile-prebuilt
+++ b/Dockerfile-prebuilt
@@ -13,4 +13,4 @@ RUN cd pyodide && make
 # this deletes the build results such that the resulting image can still be used for building pyodide
 # from source (including partial and customized builds). Due to the previous run of make, builds
 # executed with this image will be much faster than the ones executed with pyodide-env
-RUN rm -r build
+RUN rm -r pyodide/build
```
</details>

Could you please add this patch?

Another issue is the that resulting image is very large (9.8 GB), mostly due to including build folders for packages, emsdk and cpython
```
$ du -sh emsdk/ packages/ cpython/
2,3G    emsdk/
3,5G    packages/
1,1G    cpython/
```
So it's a lot, but that's also what makes partial builds fast. There is some home in reducing it in the future:
 - for emsdk once https://github.com/iodide-project/pyodide/pull/480 is resolved, clang will no longer be compiled and the folder size should be smaller (though we can probably also remove build files there now as well)
 - almost half of packages in size if scipy which would be improved with https://github.com/iodide-project/pyodide/issues/240

Could you please check that it works for you?

rth commented:
About your other comments,

> Two questions though - doesn't the build currently use all available CPUs by default? 

It doesn't I believe: PYODIDE_JOBS by defaults to 3 for sure (cf docs), and for `EMCC_CORES` we would need to check the emscripten repo/docs to be sure. Also say in CI there are 32 detected CPU but only 2 are available (limited via CPU affinity) and build systems would have trouble auto-determining that, so specyfing manually is probably safer in any case.

> Also, these vars should probably also be part of the pyodide-env image. I couldn't find the dockerfile for it (I don't think it is the one here used for circleci, or is it?)

It's the `pyodide/Docker` (the same as the one used in CI). Feel free to open another PR to improve it.

> I actually need the 0.15.1 version for my own project, since there is some kind of regression bug in 0.16 (I will try to open an issue on that in the next days)

Could you please open that issue? As I the release for 0.16 would likely happen soon. Thanks!

MischaPanch commented:
That's interesting, I built this image on a kubernetes node (from CI) on our cluster and I believe to remember that it took all 96 cores (in the 0.15 version though). Maybe I remember wrong though ^^

> 
> It doesn't I believe: PYODIDE_JOBS by defaults to 3 for sure (cf docs), and for `EMCC_CORES` we would need to check the emscripten repo/docs to be sure. Also say in CI there are 32 detected CPU but only 2 are available (limited via CPU affinity) and build systems would have trouble auto-determining that, so specyfing manually is probably safer in any case.
> 

I will have a look and make a PR in case I see something useful.
> 
> It's the `pyodide/Docker` (the same as the one used in CI). Feel free to open another PR to improve it.

It's convoluted: with the 0.16. version the pyodide files continued to work when they were called in the main thread but failed with cryptical errors in a thread managed by a web-worker. Changing the pyodide-env image back to 0.15. for transpiling python magically fixed it. I wouldn't have believed it if I hadn't seen it myself... I will try to open an issue next week but it will take some time to set up a minimal example. Probably the easiest way to do so would be to finish up my template.

> > I actually need the 0.15.1 version for my own project, since there is some kind of regression bug in 0.16 (I will try to open an issue on that in the next days)
> 
> Could you please open that issue? As I the release for 0.16 would likely happen soon. Thanks!

MischaPanch commented:
> The image [`iodide/pyodide:0.16.0b1`](https://hub.docker.com/repository/docker/iodide/pyodide) should now be available.

Done, sorry for the error in the path (typical copy-paste)
> I had to make 2 changes with respect to this PR to make it work,
> ...
> Could you please add this patch?

I also now execute the removal of built files in the same layer. This should reduce the size a bit. Originally I thought it might be useful to have it separately for faster docker builds in case the dockerfile is adjusted, but I guess size is the more important issue here. We can further improve it once the changes you mention are done, feel free to ping me for that.

> Another issue is the that resulting image is very large (9.8 GB), mostly due to including build folders for packages, emsdk and cpython

One last thing - could you also build and upload the 0.15.1 version? It can't hurt and some people might appreciate switching to the prebuilt image without having to upgrade :)

rth commented:
Thanks @MischaPanch !

> One last thing - could you also build and upload the 0.15.1 version? It also makes sense to re-upload the now smaller 0.16.0

Yes, I'll look into that and re-upload 0.16.0b2 (pre-release). There will be a 0.16.0 with the final release.

henryiii commented:
Where did this go? It seems this image is missing from the GH registry, while it was on Docker. Is there a nice path to externally compile extensions?

rth commented:
> It seems this image is missing from the GH registry, while it was on Docker

Yes, there was a CI issue when deploying the release. It was fixed since, but we would need to re-release to create the image. See https://github.com/pyodide/pyodide/issues/2163
